### PR TITLE
gh-103092: Isolate winreg

### PR DESCRIPTION
**Pull request by aisk**
* Issue: gh-103092

### CODE REVIEW

File: PC/winreg.c
Comment (arhadthedev) [inline PC/winreg.c]: A useful feature of C99:

```suggestion
    .m_base = PyModuleDef_HEAD_INIT,
    .m_name = "winreg",
    .m_doc = module_doc,
    .m_size = sizeof(winreg_state),
    .m_methods = winreg_methods,
    .m_slots = winreg_slots,
```

**Comment by Eclips4** [comment]
Hm.. CI/CD says that a some number of refleaks

**Comment by aisk** [comment]
Yes, working on it

File: Lib/test/test_winreg.py
Comment (erlend-aasland) [inline Lib/test/test_winreg.py:62]:
```suggestion


```

File: Misc/NEWS.d/next/Library/2023-04-08-00-48-40.gh-issue-103092.5EFts0.rst
Comment (erlend-aasland) [inline Misc/NEWS.d/next/Library/2023-04-08-00-48-40.gh-issue-103092.5EFts0.rst]:
```suggestion
Adapt the :mod:`winreg` extension module to :pep:`687`.
```

File: PC/winreg.c
Comment (erlend-aasland) [inline PC/winreg.c]:
```suggestion
    Py_IS_TYPE(op, ((winreg_state *)_PyModule_GetState(m))->PyHKEY_Type)
```

Comment (erlend-aasland) [reply PC/winreg.c]: Possibly needs `#include "pycore_moduleobject.h"`

File: PC/winreg.c
Comment (erlend-aasland) [inline PC/winreg.c]:
```suggestion
PyHKEY_traverseFunc(PyHKEYObject *self, visitproc visit, void *arg)
```

File: PC/winreg.c
Comment (erlend-aasland) [inline PC/winreg.c]: Align with the style used in the rest of the code base.

```suggestion
    .flags = (Py_TPFLAGS_HAVE_GC | Py_TPFLAGS_IMMUTABLETYPE |
              Py_TPFLAGS_DISALLOW_INSTANTIATION),
```

File: PC/winreg.c
Comment (erlend-aasland) [inline PC/winreg.c]: HKEY type is not subclassable, so you can revert the clinic changes and just use `Py_TYPE(self)` and then `_PyType_GetModuleState`. Ditto for `__exit__`.

File: PC/winreg.c
Comment (erlend-aasland) [inline PC/winreg.c]:
```suggestion
    PyHKEYObject *key = PyObject_GC_New(PyHKEYObject, st->PyHKEY_Type);
```

When `key` is successfully set up, you must call `PyObject_GC_Track(key)` on it.

File: PC/winreg.c
Comment (erlend-aasland) [inline PC/winreg.c]:
```suggestion
exec_module(PyObject *m)
```

File: PC/winreg.c
Comment (erlend-aasland) [reply PC/winreg.c]: You'll also need to fix up the dealloc function a la this:

```c
{
    PyTypeObject *tp = Py_TYPE(self);
    PyObject_GC_UnTrack(self);
    // Clean up (call the tp_clear() function, etc.)
    tp->tp_free((PyObject *)self);
    Py_DECREF(tp);
    return 0;
}
```

File: PC/winreg.c
Comment (aisk) [reply PC/winreg.c]: Reverted the changes and just using `Py_TYPE(self)`, but still keep using of `PyType_GetModule` instead of the suggested `_PyType_GetModuleState`.

The reason is we will passing the type instance to other functions `PyHKEY_Close` and `PyHKEY_AsHKEY`, passing a `module` as parameter is more flexible as the module level function may access more context in the module object.

I don't know is this right?

File: Lib/test/test_winreg.py
Comment (erlend-aasland) [inline Lib/test/test_winreg.py]: Nit: please join this with the line below:

```python
from test.support import cpython_only, import_helper
```

File: PC/winreg.c
Comment (erlend-aasland) [inline PC/winreg.c]:
```suggestion
    PyTypeObject *PyHKEY_Type;
```

File: PC/winreg.c
Comment (erlend-aasland) [inline PC/winreg.c]: Since we're also touching this line, please add braces around the `if` body as per PEP-7.

File: PC/winreg.c
Comment (erlend-aasland) [inline PC/winreg.c]: I'd make this an assert.

```suggestion
    assert(m != NULL);
```

File: PC/winreg.c
Comment (erlend-aasland) [inline PC/winreg.c]:
```suggestion
    assert(m != NULL);
```

File: PC/winreg.c
Comment (erlend-aasland) [inline PC/winreg.c]: Please add braces here as well.

File: PC/winreg.c
Comment (erlend-aasland) [inline PC/winreg.c:411]: `PyObject_GC_Track` will segfault if given a `NULL` pointer. I'd suggest to modify the `if` above to take care of the error path and `return NULL`, instead of letting it take care of the happy path as it does now. That will also lead to more idiomatic C code.

File: PC/winreg.c
Comment (erlend-aasland) [inline PC/winreg.c]: Please add braces here.

File: PC/winreg.c
Comment (erlend-aasland) [inline PC/winreg.c]: PEP-7 nit

```suggestion
    PyHKEYObject *op = (PyHKEYObject *)PyObject_GC_New(PyHKEYObject,
                                                       st->PyHKEY_Type);
```

Comment (erlend-aasland) [reply PC/winreg.c]: You also need to check the `PyObject_GC_New` return value, and bail if it failed.

File: PC/winreg.c
Comment (erlend-aasland) [inline PC/winreg.c]: Please add braces here.

File: PC/winreg.c
Comment (erlend-aasland) [inline PC/winreg.c]:
```suggestion
    winreg_state *state = _PyModule_GetState(module);
```

File: PC/winreg.c
Comment (erlend-aasland) [inline PC/winreg.c]:
```suggestion
    winreg_state *state = _PyModule_GetState(module);
```

File: PC/winreg.c
Comment (erlend-aasland) [inline PC/winreg.c]:
```suggestion
static int
winreg_traverse(PyObject *module, visitproc visit, void *arg)
```

File: PC/winreg.c
Comment (erlend-aasland) [inline PC/winreg.c]:
```suggestion
static int
winreg_clear(PyObject *module)
```

File: PC/winreg.c
Comment (erlend-aasland) [reply PC/winreg.c]: We don't use implicit `int`s :)

Also, style as per PEP-7

File: PC/winreg.c
Comment (erlend-aasland) [reply PC/winreg.c]: Yep, it is flexible, but for example for `winreg_ConnectRegistry`, you end up calling `PyType_GetModule` twice in the clinic code (first in `clinic_HKEY_converter`, then in `PyHKEY_FromHKEY`). Perhaps that does not matter much, wrt. performance.

File: PC/winreg.c
Comment (erlend-aasland) [inline PC/winreg.c]: Nit

```suggestion
#define PyHKEY_Check(m, op) \
```

**Review by erlend-aasland — COMMENTED** [review]
Thanks, I think we're soon ready for landing (and of course a thumbs up from Steve or someone else from the Windows team).

I left some more nit picks :)

File: PC/winreg.c
Comment (Eclips4) [inline PC/winreg.c:2110]:
```suggestion
    st->PyHKEY_Type = (PyTypeObject *)
                       PyType_FromModuleAndSpec(m, &pyhkey_type_spec, NULL);
    if (st->PyHKEY_Type == NULL)
        return -1;
```

Should we prevent this situation?

Comment (erlend-aasland) [reply PC/winreg.c:2110]: Yes, of course. Good catch!

File: PC/winreg.c
Comment (arhadthedev) [inline PC/winreg.c]: Should it be `winreg.PYHKEY` or `winreg.PyHkey`/`winreg.PyHKey`? I have no idea thus asking.

Comment (aisk) [reply PC/winreg.c]: The actual type referrence which expose to python is `winreg.HKEYType`, I think the different is quirky. For compalibility, I think we can't change the Python type name, so change the type name to `winreg.HKEYType` is more reasonable?

File: PC/winreg.c
Comment (erlend-aasland) [inline PC/winreg.c:478]: Please add error handling here and bail if `op == NULL`.

File: PC/winreg.c
Comment (erlend-aasland) [reply PC/winreg.c]: @aisk is correct in that we expose this as `HKEYType`, but we should keep `__name__` intact, so please use the same name as before: `"PyHKEY"`. Adding the module prefix should be ok, though.

```suggestion
    .name = "winreg.PyHKEY",
```

Comment (aisk) [reply PC/winreg.c]: Oh I just realized it ...

File: PC/winreg.c
Comment (erlend-aasland) [inline PC/winreg.c:366]: `PyHKEY_NumberMethods` is now unused; it can be removed.

File: PC/winreg.c
Comment (erlend-aasland) [inline PC/winreg.c:370]:
```suggestion

    // Number protocol
    {Py_nb_add, PyHKEY_binaryFailureFunc},
```

File: PC/winreg.c
Comment (erlend-aasland) [inline PC/winreg.c]:
```suggestion
```

**Review by erlend-aasland — APPROVED** [review]

**Comment by erlend-aasland** [comment]
Thanks, @aisk!

Let's wait for Steve's thumbs up before we land this.

**Comment by aisk** [comment]
Great thanks for helping and review this! @erlend-aasland

**Comment by zooba** [comment]
Seems okay to me, though I'm not really up on these conversions. If all the tests (esp. refleak tests) are still passing, it's presumably fine.

**Pull request merged**
